From 57fcb669d0e29f33a2b490e2d3c894714966aee6 Mon Sep 17 00:00:00 2001
From: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
Date: Wed, 27 Jan 2016 11:19:54 +0530
Subject: [PATCH 04/24] usb_pd: ignore timer faults in default and error
 recovery states.

Any timer faults when pe is in PE_STATE_NONE or ERROR_RECOVEY
should be ignore and pe should retain same state.

Change-Id: I39ec68115548d0d8adca175fbe5569d93511e665
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-18527
Signed-off-by: Venkataramana Kotakonda <venkataramana.kotakonda@intel.com>
Reviewed-on: https://android.intel.com:443/462779
---
 drivers/usb/typec/pd/policy_engine.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/drivers/usb/typec/pd/policy_engine.c b/drivers/usb/typec/pd/policy_engine.c
index b70d15b..8fbbd42e 100644
--- a/drivers/usb/typec/pd/policy_engine.c
+++ b/drivers/usb/typec/pd/policy_engine.c
@@ -1347,10 +1347,17 @@ static void pe_timer_expire_worker(struct work_struct *work)
 	enum pe_timers type = cur_timer->timer_type;
 
 	mutex_lock(&pe->pe_lock);
+	if (pe->cur_state == PE_ERROR_RECOVERY
+		|| pe->cur_state == PE_STATE_NONE) {
+		log_info("Ignore timer expiries in state=%d",
+				pe->cur_state);
+		mutex_unlock(&pe->pe_lock);
+		return;
+	}
 	log_dbg("%s expiration handling!!!",
 			timer_to_str(type));
-	switch (type) {
 
+	switch (type) {
 	case SENDER_RESPONSE_TIMER:
 		if (pe->cur_state == PE_SRC_SEND_CAPABILITIES) {
 			if (pe->is_gcrc_received)
@@ -2423,11 +2430,12 @@ pe_process_state_pe_error_recovery(struct policy_engine *pe)
 static void
 pe_process_state_pe_state_none(struct policy_engine *pe)
 {
-	pe_do_complete_reset(pe);
-	/* VBUS Off */
-	devpolicy_set_vbus_state(pe->p.dpm, false);
+	pe_deactivate_all_timers(pe);
 	/*VCONN off */
 	devpolicy_set_vconn_state(pe->p.dpm, VCONN_NONE);
+	/* VBUS Off */
+	devpolicy_set_vbus_state(pe->p.dpm, false);
+	pe_do_complete_reset(pe);
 	pe_set_data_role(pe, DATA_ROLE_NONE);
 	pe_set_power_role(pe, POWER_ROLE_NONE);
 	if (pe->prev_state == PE_ERROR_RECOVERY)
-- 
1.9.1

